### PR TITLE
ChannelsLast3d support is_contiguous, contiguous, suggest_memory_format, caching

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33033 ChannelsLast3d support is_contiguous, contiguous, suggest_memory_format, caching**

- Support the following features for Channels Last 3d
  is_contiguous()
  contiguous()
  suggest_memory_format
  empty()
  empty_like()
  caching.
  operators.
  etc. (related functions that depends on above)
- A bug fix
  see is_channels_last_strides() for details

Allow 5d channels last memory format. (dim() == 5 only for this PR) 
Added all test cases.

Differential Revision: [D19759661](https://our.internmc.facebook.com/intern/diff/D19759661)

### CODE REVIEW

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Tiny changes needed, also please benchmark new code.

File: c10/core/TensorImpl.h [hunk 1]
```diff
@@ -1482,14 +1482,23 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
...
     is_contiguous_ = compute_contiguous();
-    is_channels_last_contiguous_ = compute_channels_last_contiguous(MemoryFormat::ChannelsLast);
-    is_channels_last_3d_contiguous_ = compute_channels_last_contiguous(MemoryFormat::ChannelsLast3d);
-    // is_channels_last_ and is_channels_last_3d_ are suggested memory_format.
-    // Being channels_last_contiguous doesn't necessarily mean the tensor is
-    // strided like channels_last: for strides on channel dimension could suggest
-    // desired memory_layout, but it doesn't affect memory storage
-    is_channels_last_ = compute_strides_like_channels_last(MemoryFormat::ChannelsLast);
-    is_channels_last_3d_ = (!is_channels_last_ && compute_strides_like_channels_last(MemoryFormat::ChannelsLast3d));
+    if (!is_contiguous_) {
```
Comment (VitalyFedyunin) [inline c10/core/TensorImpl.h]: This is incorrect, tensor can be contiguous and channels last simulteniously, for example NC11 tensors.

Comment (glaringlee) [reply c10/core/TensorImpl.h]: I see. Want to speed up this part, will send a new PR later.

**Comment by glaringlee** [comment]
d663f66 is to rebase and fix CI test. Will discuss the logic later

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by glaringlee** [comment]
checkin a new version to resolve merge conflict

**Comment by glaringlee** [comment]
@ezyang 
I noticed these and hit the assert for sure. It was 30 words and I changed it to 31 words. I did discuss this size increasing offline but did not leave any comments in this PR.  I was wondering the meaning of `miscellaneous bitfield` but I did not get the purpose that we want it be 8 bytes, i've thought that is something reserved since otherwise, 4bytes can also be used but we did not use it (we have 7 booleans before this change)

People do read this note. But I think there are 2 things that can be more clear.
1.  Comment the meaning of `miscellaneous bitfield`.  Now, I guess it should be `You should pack all miscellaneous boolean fields in it` (after changing the code to use bitfield)
2.  The process of handling the size increases.  For eg, asking review for specific POC? (or passing  https://fburl.com/q5enpv98 is fine?)

**Comment by ezyang** [comment]
> The process of handling the size increases. For eg, asking review for specific POC? (or passing https://fburl.com/q5enpv98 is fine?)

This is tricky. I think the basic sniff test is "am I increasing the size for a good reason, or is there an easy way to prevent the size from increasing." Bools to bitmap is a good example of this. Forcing people to run a particular workflow is not great, since workflows will become obsolete over time.

**Pull request closed without merging**